### DISCUSSION TITLE
Feature request: Hooks on saving a bookmark

### DISCUSSION THREAD

**Discussion by lenormf** [go-shiori/shiori#826] [Feature requests]
Hi!

I think Shiori would be more flexible/extensible if it had a hook mechanism upon saving a bookmark, new or otherwise.

Ideally, the Shiori server would call a script (whose path would be set in the configuration) right after it has set the various metadata for a given bookmark, and override those automatically. How that would be handled is to be determined (maybe allow the script to call the Shiori API by passing it a temporary token and the ID/metadata of the bookmark, as arguments?).

I have few use cases:

1. Shiori won’t archive Youtube pages, nor any other pages that require JavaScript, so I could implement a script that handles that and feeds Shiori with the right data.
2. I would like to grab whatever _contents_ Shiori has detected on the page and index it, in another database (e.g. [Sonic](https://github.com/valeriansaliou/sonic)).
3. Anything else that requires logic, e.g.

    - _Auto Tagging Rules_ #460
    - _Automatically fill in fields on create_ #538

I’m willing to get a PR going once the details are ironed-out, if you agree with the feature proposal.

Thank you!

**Comment by Monirzadeh** [comment]
i thinking about a plugin system for Shiori. this keeps Shiori simpler to maintain in long term. and user can add this kind of feature too. 
what do you think? can you send a PR to impediment something like this?

> Shiori won’t archive Youtube pages, nor any other pages that require JavaScript, so I could implement a script that handles that and feeds Shiori with the right data.

maybe we should solve this in [obelisk ](https://github.com/go-shiori/obelisk) that Shiori should move to that instead of warc or [go-readability](https://github.com/go-shiori/go-readability) for better handling readable version.

**Reply by lenormf** [reply]
Handling pages generated via Javascript was only one example, I also would like to handle direct links to files (e.g. ebooks) somehow.

However this request is not about how to handle those cases directly, but rather to provide a mechanism in Shiori that would allow interfacing the server with any tool of the user’s choosing, so that they (we) can implement a solution independently.

**Comment by fmartingr** [comment]
Ideally I was thinking more of letting using create webhooks on actions, URLs that would be called on bookmark creation for example, and those webhooks perform the action or maybe we could support some directly ourselves.

The Plugin API may be possible but it would need to be maintained too, and I haven't thought on how to properly implement it and how contributors could create their own plugins.

I think letting Shiori run arbitrary scripts in a bad idea, even if those are set up by the instance admin/local user.

**Reply by lenormf** [reply]
Could you explain what webhooks would look like?

It sounds to be what I am proposing, but I don’t see how I could use them, from a configuration or UI-setting point of view.

**Reply by fmartingr** [reply]
The main idea around this would be to allow users to setup outgoing webhooks from the configuration UI (and potentially from the CLI, but not 100% sold about that). 
- Create a new webhook providing an URL and a method (potentially auth, etc)
- When a bookmark is saved (created, updated, deleted) the request will be sent with a defined payload specifying what happened in the event and the bookmark details
- The end system can parse the data and act accordingly

This is a very rough idea. Before this we need a robust background job system and potentially proper multi-user support.
